lib/Dialects/VectorExt/Transform/VectorMaskingUtils.cpp

```diff
@@ -0,0 +1,156 @@
+  OpBuilder::InsertionGuard guard(builder);
+  builder.setInsertionPointToStart(&blockToPredicate);
+  Value predicate = createPredicate(builder);
+
+  // Generate the vector.predicate operation and move 'opsToMove' within its
+  // truePredicateRegion. We have to rewire the def-use chain for those
+  // definitions within the range that have external uses. Those uses are
+  // rewired to the results of the vector.predicate.
+  Location loc = op->getLoc();
+  auto vecPredOp = builder.create<PredicateOp>(loc, defTypes, predicate);
+  assert(defsWithUsesOutside.size() == vecPredOp.getNumResults() &&
+         "Expected same size");
+  for (auto &en : llvm::enumerate(defsWithUsesOutside))
+    en.value().replaceAllUsesWith(vecPredOp.getResult(en.index()));
+
+  Operation *truePredTerminator =
+      &vecPredOp.truePredicateRegion().front().back();
+  moveOperationsBefore(opsToMove, truePredTerminator);
+
+  // The existing terminator of TruePredicateRegion doesn't yield any value.
+  // Replace it with a new terminator that returns the definitions with uses
+  // outside that we just moved.
+  if (vecPredOp.getNumResults() > 0) {
+    builder.setInsertionPoint(truePredTerminator);
+    builder.create<vector_ext::YieldOp>(loc, defsWithUsesOutside);
+    truePredTerminator->erase();
+  }
+
+  return vecPredOp;
+}
+
+/// Utility that predicates the body a tiled loop with a vector.predicate
+/// operation. The vectorization factor used for predication is assumed to be
+/// the step of the tiled loop.
+LogicalResult mlir::vector_ext::predicateTiledLoop(TiledLoopOp loopOp) {
+  if (loopOp.lowerBound().size() > 1)
+    // TODO: Support multi-dim tiled loops.
+    return failure();
+
+  // Retrieve vectorization factor from the step of the tiled loop.
+  Value step = loopOp.step()[0];
+  auto maybeVecFactor = getConstantIntValue(step);
+  if (!maybeVecFactor)
+    return failure();
+  int64_t vecFactor = *maybeVecFactor;
+
+  auto createPredicate = [&](OpBuilder &builder) -> Value {
+    // Generate the predicate to used by the vector.predicate operation:
+    //   %min = affine.min affine_map<(d0)[s0] -> (vecFactor, -d0 +
+    //   s0)>(%iv)[%ub] %vec_pred = vector.create_mask %min : vector<8xi1>
+    // TODO: This is probably not the most efficient way to generate the vector
+    // predicate. Consider using the vector IV.
+    AffineExpr i, j;
+    bindDims(loopOp.getContext(), i, j);
+    SmallVector<AffineMap, 4> maps = AffineMap::inferFromExprList(
+        ExprList{{builder.getAffineConstantExpr(vecFactor), i - j}});
+
+    Location loc = loopOp.getLoc();
+    auto minOp = builder.create<AffineMinOp>(
+        loc, loopOp.step()[0].getType(), maps[0],
+        ValueRange{loopOp.upperBound()[0], loopOp.getInductionVars()[0]});
+
+    auto maskType = VectorType::get({vecFactor}, builder.getI1Type());
+    return builder.create<CreateMaskOp>(loc, maskType, ValueRange{minOp});
+  };
+
+  OpBuilder builder(loopOp);
```

Comment:
Builder creation within the op will make it hard to compose, please pass from above.